README.md
# Zepyhr SPi flash driver for EN25QH32B-104HIP2C flash chip

This repository contains driver and tests for interfacing with
EN25QH32B-104HIP2C via the zephyr flash API.

This drive is compatible with the following flash chips:

- `EN25QH32B` with JEDEC id `[1c 70 16]`
- `IS25WP032D` with JEDEC id `[9D 70 16]`
- `W25Q32JVSSIQ` with JEDEC id `[EF 40 16]`
- `AT25SL321` with JEDEC id `[1F 42 16]`

## Setup

1. To your `west.yml` add the irnas remote to the `remotes` section:

   ```yaml
   - name: irnas
     url-base: https://github.com/irnas
   ```

2. Then in the `projects` section add at the bottom:

   ```yaml
   - name: zephyr-spi-flash-en25-driver
     repo-path: zephyr-spi-flash-en25-driver
     path: irnas/zephyr-spi-flash-en25-driver
     remote: irnas
     revision: <revision>
   ```

3. Add the flash DTS entry to your board definition or overlay file. For
   example:

```dts

&spi0 {
   // ...

    en25qh32b: en25qh32b@0 {
        reg = <0>;
        label = "EN25QH32B";
        status = "okay";
        compatible = "mxicy,en25";

        jedec-id = [ 1c 70 16  ];  // EN25
        size = <(4194304 * 8)>;

        write-sector-size = <256>;
        erase-full-block-size = <65536>;
        erase-half-block-size = <32768>;
        erase-sector-size = <4096>;

        spi-max-frequency = <4000000>;

        enter-dpd-delay = <30>;
        exit-dpd-delay = <30>;

        wp-gpios = <&gpio0 22 0>;
        hold-gpios = <&gpio0 23 0>;
    };
};

```

## External mutex

This driver can be used on multiple MCUs to use the same SPI flash peripheral.
To achieve this, a GPIO line must be shared between the MCUs, the flash chip
must be the only peripheral on this SPI bus and both MCUs must be using this
driver to communicate with the external flash. To enable this feature, add
`ext-mutex-gpios`, `ext-mutex-role` and `spi-clk-gpios` to the DTS flash
definition.

For two MCUs, set one `ext-mutex-role` to `master` and one to `slave`. For more
than 2 MCUs, set one to `master` and all others to `slave`.

For example:

```dts

&spi0 {
   // ...

    en25qh32b: en25qh32b@0 {
        // ...

        ext-mutex-gpios =  <&gpio0 24 (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>;
        ext-mutex-role = "slave";
        spi-clk-gpios = <&gpio0 28 0>;  // must be same as the pinctrl entry of the enclosing spi peripheral
    };
};

```

The setting `SPI_FLASH_EN25_EXTERNAL_MUTEX_TIMEOUT` can also be configured to
specify the amount of time a MCU is willing to wait for the SPI lock to be
released.

## Tests

1. Navigate to `./tests/flash_read_write`
2. Build for one of the boards with supplied overlay, of make your own. Use
   `west build -b nrf52840dk_nrf52840` for example.
3. flash with `west flash`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[IRNAS's naming guidelines](https://github.com/IRNAS/irnas-core/blob/master/GITHUB_NAMING_GUIDELINES.md).

## [Unreleased]

## [3.4.0] - 2023-09-01

### Added

-   Migrate to trunk-based development workflow

### Changed

-   Removed half region write test from test `flash_read_write`
-   Updated the driver to NCS 2.4.0.
-   Added west.yml file
-   Replace irnas vendor prefix with mxicy

## [3.3.0] - 2023-06-22

### Added

-   Read one byte after init to enable low power operation.

### Fixed

-   Fix mutex handling during driver initialization.
-   Configure CS pin to disconnected pullup on mutex release

## [3.2.0] - 2023-03-07

### Changed

-   Update driver for NCS 2.2.

## [3.1.0] - 2023-01-25

### Added

-   SPI external mutex functionality.

## [3.0.0] - 2022-09-26

### Changed

-   Rework how sector sizes are specified in DTS.

## [2.1.0] - 2022-07-07

### Changed

-   Remove unnecessary board overlays form tests.
-   Migrate code to run on NCS 2.0

## [2.0.5] - 2022-05-03

### Added

-   Added Kconfig option `CONFIG_SPI_FLASH_EN25_JEDEC_CHECK_AT_INIT` with default
    `n`. If set to `y`, the jedec id check will be performed during driver init
    and will not init the glash device if it fails.

### Fixed

-   Fix errors in the test suite.
-   Fix error handling in `relese_ext_mutex`.

## [2.0.4] - 2022-04-19

### Fixed

-   Remove double call to acquire in `spi_flash_en25_pm_control`.

## [2.0.3] - 2022-03-23

### Fixed

-   Replace deprecated `CONFIG_DEVICE_POWER_MANAGEMENT` with `CONFIG_PM_DEVICE`

## [2.0.2] - 2022-02-17

### Fixed

-   Fixed spi_flash_en25_pm_control not behaving correctly and not returning any
    value at the end of the function.

## [2.0.1] - 2022-02-17

### Added

-   Configurable wait for ready timeout via the
    `CONFIG_SPI_FLASH_EN25_READY_TIMEOUT` KConfig option.

### Changed

-   Removed jedec id check in erase function.

## [2.0.0] - 2022-01-20

### Added

-   Support for `WP` and `HOLD` pins in driver and device tree bindings.

### Changed

-   Power management control due to API changes in NCS 1.8. This means that this
    version of the driver is not compatible with older NCS versions anymore.
-   Driver does not fail at init if the wrong jedec id is read at init, but it
    will report the error over logging.
-   Driver will perform jedec id check at the start of flash erase and will throw
    \-ENODEV in case of error.
-   Infinite loop in `wait_until_ready` has been replaced with the 10 second
    timeout.

### Extra

-   Above changes related to jedec check id should be removed when the lr1110
    driver is rewritten with zephyr API or when a way to move flash driver init
    after the lr1110 driver init is found.

[Unreleased]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v3.4.0...HEAD

[3.4.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v3.3.0...v3.4.0

[3.3.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v3.2.0...v3.3.0

[3.2.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v3.1.0...v3.2.0

[3.1.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v3.0.0...v3.1.0

[3.0.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare//v2.1.0...v3.0.0

[2.1.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v2.0.5.../v2.1.0

[2.0.5]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v2.0.4.../v2.0.5

[2.0.4]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v2.0.3.../v2.0.4

[2.0.3]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v2.0.2.../v2.0.3

[2.0.2]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v2.0.1.../v2.0.2

[2.0.1]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v2.0.0.../v2.0.1

[2.0.0]: https://github.com/IRNAS/zephyr-spi-flash-en25-driver/compare/v1.1.1.../v2.0.0


tests/flash_read_write/src/main.c
#include <zephyr/devicetree.h>
#include <zephyr/drivers/flash.h>
#include <zephyr/drivers/gpio.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log_ctrl.h>
#include <zephyr/pm/device.h>
#include <zephyr/storage/flash_map.h>
#include <zephyr/ztest.h>

#define CHIP_SIZE_BITS	  DT_PROP(DT_NODELABEL(en25qh32b), size)
#define ERASE_SECTOR_SIZE DT_PROP(DT_NODELABEL(en25qh32b), erase_sector_size)

/* Since we erase the test region, it's offset must be a multiple of the erase-sector-size */
#define TEST_REGION_OFFSET (ERASE_SECTOR_SIZE * 4)
/* The size of test region must also be a multiple of the erase-sector-size, for the same reason */
#define TEST_REGION_SIZE   (ERASE_SECTOR_SIZE * 2)
#define ERASE_BUF_SIZE	   ERASE_SECTOR_SIZE

/* Define stattically so not a lot of additional stack space is required */
static uint8_t write_buf[TEST_REGION_SIZE];
static uint8_t read_buf[TEST_REGION_SIZE];
static uint8_t erase_check_buf[ERASE_BUF_SIZE];

const struct device *flash_dev = DEVICE_DT_GET(DT_NODELABEL(en25qh32b));

ZTEST_SUITE(flash_test_suite, NULL, NULL, NULL, NULL, NULL);

ZTEST(flash_test_suite, test_full_erase_full_read_write)
{
	int err;
	uint8_t data;
	struct flash_pages_info pages_info;
	size_t page_count, chip_size;

	page_count = flash_get_page_count(flash_dev);
	(void)flash_get_page_info_by_idx(flash_dev, 0, &pages_info);
	chip_size = page_count * pages_info.size;
	err = flash_read(flash_dev, TEST_REGION_OFFSET, &data, 1);

	/* Chip size that we get from multiplication is in bytes,
	 * but CHIP_SIZE_BITS that we get from device tree macro is in bits*/
	zassert_equal((chip_size * 8), CHIP_SIZE_BITS, "Chip sizes are not equal");
	zassert_equal(pages_info.size, ERASE_SECTOR_SIZE, "Page sizes are not equal");
	zassert_equal(err, 0, "Flash read failed");

	printk(" INFO - Starting full flash erase, this will take around 15 "
	       "seconds...\n");
	err = flash_erase(flash_dev, 0, chip_size);
	zassert_equal(err, 0, "Full flash erase failed");

	printk(" INFO - Checking that everything was erased, this will take some "
	       "time...\n");
	for (int i = 0; i < page_count; i++) {
		err = flash_read(flash_dev, ERASE_BUF_SIZE * i, erase_check_buf, ERASE_BUF_SIZE);
		zassert_equal(err, 0, "Flash read failed at i: %d", i);
		for (int j = 0; j < ERASE_BUF_SIZE; j++) {
			zassert_equal(erase_check_buf[j], 0xFF,
				      "ERROR at erase_check_buf[%d]: expected 0x%02X, got "
				      "0x%02X\n,\n i is: %d",
				      j, 0xFF, erase_check_buf[j], i);
		}
	}

	printk(" INFO - Writing entire flash with dummy values, this will take some "
	       "time...\n");
	for (int i = 0; i < ERASE_BUF_SIZE; ++i) {
		erase_check_buf[i] = i;
	}
	for (int i = 0; i < page_count; i++) {
		err = flash_write(flash_dev, ERASE_BUF_SIZE * i, erase_check_buf, ERASE_BUF_SIZE);
		zassert_equal(err, 0, "Flash write failed at i: %d", i);
	}

	/* Empty buffer just in case*/
	for (int i = 0; i < ERASE_BUF_SIZE; ++i) {
		erase_check_buf[i] = 0;
	}

	printk(" INFO - Reading entire flash, this will take some time...\n");
	for (int i = 0; i < page_count; i++) {
		err = flash_read(flash_dev, ERASE_BUF_SIZE * i, erase_check_buf, ERASE_BUF_SIZE);
		zassert_equal(err, 0, "Flash read failed at i: %d", i);
		for (int j = 0; j < ERASE_BUF_SIZE; j++) {
			zassert_equal(erase_check_buf[j], (uint8_t)j,
				      "ERROR at erase_check_buf[%d]: expected 0x%02X, got "
				      "0x%02X\n,\n i is: %d",
				      j, j, erase_check_buf[j], i);
		}
	}
}

ZTEST(flash_test_suite, test_erase_read_write)
{
	int err;
	uint8_t data;
	struct flash_pages_info pages_info;
	size_t page_count, chip_size;

	page_count = flash_get_page_count(flash_dev);
	(void)flash_get_page_info_by_idx(flash_dev, 0, &pages_info);
	chip_size = page_count * pages_info.size;
	err = flash_read(flash_dev, TEST_REGION_OFFSET, &data, 1);

	/* Chip size that we get from multiplication is in bytes,
	 * but CHIP_SIZE_BITS that we get form device tree macro is in bits*/
	zassert_equal((chip_size * 8), CHIP_SIZE_BITS, "Chip sizes are not equal");
	zassert_equal(pages_info.size, ERASE_SECTOR_SIZE, "Page sizes are not equal");
	zassert_equal(err, 0, "Flash read failed");

	++data;
	for (int i = 0; i < TEST_REGION_SIZE; ++i) {
		write_buf[i] = (uint8_t)(data + i);
	}

	err = flash_erase(flash_dev, TEST_REGION_OFFSET, TEST_REGION_SIZE);
	zassert_equal(err, 0, "Flash region erase failed");

	err = flash_read(flash_dev, TEST_REGION_OFFSET, read_buf, TEST_REGION_SIZE);
	zassert_equal(err, 0, "Flash read failed");

	for (int i = 0; i < TEST_REGION_SIZE; ++i) {
		zassert_equal(read_buf[i], 0xFF,
			      "ERROR at read_buf[%d]: expected 0x%02X, got 0x%02X\n", i, 0xFF,
			      read_buf[i]);
	}

	err = flash_write(flash_dev, TEST_REGION_OFFSET, write_buf, TEST_REGION_SIZE);
	zassert_equal(err, 0, "Flash write failed");

	err = flash_read(flash_dev, TEST_REGION_OFFSET, read_buf, TEST_REGION_SIZE);
	zassert_equal(err, 0, "Flash read failed");

	for (int i = 0; i < TEST_REGION_SIZE; ++i) {
		zassert_equal(read_buf[i], write_buf[i],
			      "ERROR at read_buf[%d]: expected 0x%02X, got 0x%02X\n", i,
			      write_buf[i], read_buf[i]);
	}
}

#define TEST_AREA_MAX DT_PROP(DT_NODELABEL(en25qh32b), size)
#define EXPECTED_SIZE 1024
#define CANARY	      0xff

static uint8_t __aligned(4) expected[EXPECTED_SIZE];

ZTEST(flash_test_suite, test_setup1)
{
	int rc;
	const struct device *flash_dev;
	struct flash_pages_info page_info;

	flash_dev = DEVICE_DT_GET(DT_NODELABEL(en25qh32b));
	const struct flash_parameters *flash_parameters = flash_get_parameters(flash_dev);

	/* For tests purposes use page (in nrf_qspi_nor page = 64 kB) */
	flash_get_page_info_by_offs(flash_dev, TEST_REGION_OFFSET, &page_info);

	/* Check if test region is not empty */
	uint8_t buf[EXPECTED_SIZE];

	rc = flash_read(flash_dev, TEST_REGION_OFFSET, buf, EXPECTED_SIZE);
	zassert_equal(rc, 0, "Cannot read flash");

	/* Fill test buffer with random data */
	for (int i = 0; i < EXPECTED_SIZE; i++) {
		expected[i] = i;
	}

	/* Check if tested region fits in flash */
	zassert_true((TEST_REGION_OFFSET + EXPECTED_SIZE) < TEST_AREA_MAX,
		     "Test area exceeds flash size");

	/* Check if flash is cleared */
	bool is_buf_clear = true;

	for (off_t i = 0; i < EXPECTED_SIZE; i++) {
		if (buf[i] != flash_parameters->erase_value) {
			is_buf_clear = false;
			break;
		}
	}

	if (!is_buf_clear) {
		/* erase page */
		rc = flash_erase(flash_dev, page_info.start_offset, page_info.size);
		zassert_equal(rc, 0, "Flash memory not properly erased");
	}
}

ZTEST(flash_test_suite, test_read_unaligned_address)
{
	int rc;
	uint8_t buf[EXPECTED_SIZE];
	const struct device *flash_dev;
	struct flash_pages_info page_info;

	flash_dev = DEVICE_DT_GET(DT_NODELABEL(en25qh32b));

	flash_get_page_info_by_offs(flash_dev, TEST_REGION_OFFSET, &page_info);

	rc = flash_write(flash_dev, page_info.start_offset, expected, EXPECTED_SIZE);
	zassert_equal(rc, 0, "Cannot write to flash");

	/* read buffer length*/
	for (off_t len = 0; len < 25; len++) {
		/* address offset */
		for (off_t ad_o = 0; ad_o < 4; ad_o++) {
			/* buffer offset; leave space for buffer guard */
			for (off_t buf_o = 1; buf_o < 5; buf_o++) {
				/* buffer overflow protection */
				buf[buf_o - 1] = CANARY;
				buf[buf_o + len] = CANARY;
				memset(buf + buf_o, 0, len);
				rc = flash_read(flash_dev, page_info.start_offset + ad_o,
						buf + buf_o, len);
				zassert_equal(rc, 0, "Cannot read flash");
				zassert_equal(memcmp(buf + buf_o, expected + ad_o, len), 0,
					      "Flash read failed at len=%d, "
					      "ad_o=%d, buf_o=%d",
					      len, ad_o, buf_o);
				/* check buffer guards */
				zassert_equal(buf[buf_o - 1], CANARY,
					      "Buffer underflow at len=%d, "
					      "ad_o=%d, buf_o=%d",
					      len, ad_o, buf_o);
				zassert_equal(buf[buf_o + len], CANARY,
					      "Buffer overflow at len=%d, "
					      "ad_o=%d, buf_o=%d",
					      len, ad_o, buf_o);
			}
		}
	}
}

ZTEST(flash_test_suite, test_low_power)
{
#if IS_ENABLED(CONFIG_PM_DEVICE)
	printk("Putting the flash device into low power state...\n");
	int err = pm_device_action_run(flash_dev, PM_DEVICE_ACTION_SUSPEND);
	zassert_equal(err, 0, "Setting low power mode failed");
#endif
}


src/main.c
#include <zephyr.h>
#include <drivers/flash.h>
#include <logging/log_ctrl.h>

#include <zephyr.h>
#include <ztest.h>
#include <drivers/flash.h>
#include <drivers/gpio.h>
#include <devicetree.h>
#include <storage/flash_map.h>

#define FLASH_DEVICE DT_LABEL(DT_NODELABEL(flash42))
#define CHIP_SIZE DT_PROP(DT_NODELABEL(flash42), size)
#define PAGE_SIZE DT_PROP(DT_NODELABEL(flash42), page_size)

#define TEST_REGION_OFFSET  0xFE00
#define TEST_REGION_SIZE    0x1200

/* For some reason this has to be declared statically, if declared inside a
 * function, an error is throwed. */
static uint8_t write_buf[TEST_REGION_SIZE];
static uint8_t read_buf[TEST_REGION_SIZE];

static void test_get_binding()
{
	const struct device *flash_dev;
	flash_dev = device_get_binding(FLASH_DEVICE);
    zassert_not_null(flash_dev, "Returned device is null");
    
    if (strcmp(CONFIG_BOARD, "nrf9160dk_nrf9160") == 0)
    {
      const struct device *gpio_dev;
      gpio_dev = device_get_binding("GPIO_0");
      gpio_pin_configure(gpio_dev, 16, GPIO_OUTPUT);  // LR1110 reset
      gpio_pin_configure(gpio_dev, 18, GPIO_OUTPUT);  // LR1110 nss
      gpio_pin_configure(gpio_dev, 19, GPIO_OUTPUT);  // LR1110 pwr en
      gpio_pin_set(gpio_dev, 16, 1);
      gpio_pin_set(gpio_dev, 18, 1);
      gpio_pin_set(gpio_dev, 19, 1);
    }
}

#define ERASE_BUF_SIZE PAGE_SIZE      /* This should be equal to page size*/
static uint8_t erase_check_buf[ERASE_BUF_SIZE];

static void test_full_erase_full_read_write()
{
	const struct device *flash_dev;
	int i,j;
	int err;
	uint8_t data;
	struct flash_pages_info pages_info;
	size_t page_count, chip_size;


	flash_dev = device_get_binding(FLASH_DEVICE);
    zassert_not_null(flash_dev, "Returned device is null");

    page_count = flash_get_page_count(flash_dev);
    (void)flash_get_page_info_by_idx(flash_dev, 0, &pages_info);
    chip_size = page_count * pages_info.size;
    err = flash_read(flash_dev, TEST_REGION_OFFSET, &data, 1);

    /* Chip size that we get from multiplication is in bytes,
     * but CHIP_SIZE that we get from device tree macro is in bits*/
    zassert_equal((chip_size*8), CHIP_SIZE, "Chip sizes are not equal");
    zassert_equal(pages_info.size, PAGE_SIZE, "Page sizes are not equal");
    zassert_equal(err, 0, "Flash read failed");

	printk(" INFO - Starting full flash erase, this will take around 15 seconds...\n");
	err = flash_erase(flash_dev, 0, chip_size);
	printk(" INFO - Starting full flash erase, this will take around 15 seconds...\n");
    zassert_equal(err, 0, "Full flash erase failed");

	printk(" INFO - Checking that everything was erased, this will take some time...\n");
    for (i = 0; i < page_count; i++) {
        err = flash_read(flash_dev, ERASE_BUF_SIZE * i, erase_check_buf, ERASE_BUF_SIZE);
        zassert_equal(err, 0, "Flash read failed at i: %d", i);
        for (j = 0; j < ERASE_BUF_SIZE; j++) {
            zassert_equal(erase_check_buf[j], 0xFF, 
                    "ERROR at erase_check_buf[%d]: expected 0x%02X, got 0x%02X\n,\n i is: %d", 
                    j, 0xFF, erase_check_buf[j], i);
        }
    }


	printk(" INFO - Writing entire flash with dummy values, this will take some time...\n");
    for (i = 0; i < ERASE_BUF_SIZE; ++i) {
        erase_check_buf[i] = i;
    }
    for (i = 0; i < page_count; i++) {
        err = flash_write(flash_dev, ERASE_BUF_SIZE * i, erase_check_buf, ERASE_BUF_SIZE);
        zassert_equal(err, 0, "Flash read failed at i: %d", i);
    }

    /* Erase just in case*/
    for (i = 0; i < ERASE_BUF_SIZE; ++i) {
        erase_check_buf[i] = 0;
    }

	printk(" INFO - Reading entire flash, this will take some time...\n");
    for (i = 0; i < page_count; i++) {
        err = flash_read(flash_dev, ERASE_BUF_SIZE * i, erase_check_buf, ERASE_BUF_SIZE);
        zassert_equal(err, 0, "Flash read failed at i: %d", i);
        for (j = 0; j < ERASE_BUF_SIZE; j++) {
            zassert_equal(erase_check_buf[j], j, 
                    "ERROR at erase_check_buf[%d]: expected 0x%02X, got 0x%02X\n,\n i is: %d", 
                    j, j, erase_check_buf[j], i);
        }
    }
}

static void test_erase_read_write()
{
	const struct device *flash_dev;
	int i;
	int err;
	uint8_t data;
	struct flash_pages_info pages_info;
	size_t page_count, chip_size;


	flash_dev = device_get_binding(FLASH_DEVICE);
    zassert_not_null(flash_dev, "Returned device is null");

    page_count = flash_get_page_count(flash_dev);
    (void)flash_get_page_info_by_idx(flash_dev, 0, &pages_info);
    chip_size = page_count * pages_info.size;
    err = flash_read(flash_dev, TEST_REGION_OFFSET, &data, 1);

    /* Chip size that we get from multiplication is in bytes,
     * but CHIP_SIZE that we get form device tree macro is in bits*/
    zassert_equal((chip_size*8), CHIP_SIZE, "Chip sizes are not equal");
    zassert_equal(pages_info.size, PAGE_SIZE, "Page sizes are not equal");
    zassert_equal(err, 0, "Flash read failed");

    ++data;
    for (i = 0; i < TEST_REGION_SIZE; ++i) {
        write_buf[i] = (uint8_t)(data + i);
    }

    err = flash_erase(flash_dev, TEST_REGION_OFFSET, TEST_REGION_SIZE);
    zassert_equal(err, 0, "Flash region erase failed");


    err = flash_read(flash_dev, TEST_REGION_OFFSET, read_buf, TEST_REGION_SIZE);
    zassert_equal(err, 0, "Flash read failed");

	for (i = 0; i < TEST_REGION_SIZE; ++i) {
        zassert_equal(read_buf[i], 0xFF, 
                "ERROR at read_buf[%d]: expected 0x%02X, got 0x%02X\n", 
                i, 0xFF, read_buf[i]);
	}

    err = flash_write(flash_dev, TEST_REGION_OFFSET, write_buf, TEST_REGION_SIZE);
    zassert_equal(err, 0, "Flash write failed");

    err = flash_write(flash_dev, TEST_REGION_OFFSET + TEST_REGION_SIZE/2, 
            &write_buf[TEST_REGION_SIZE/2], TEST_REGION_SIZE/2);
    zassert_equal(err, 0, "Flash write failed");

    err = flash_read(flash_dev, TEST_REGION_OFFSET, read_buf, TEST_REGION_SIZE);
    zassert_equal(err, 0, "Flash read failed");

	for (i = 0; i < TEST_REGION_SIZE; ++i) {
        zassert_equal(read_buf[i], write_buf[i], 
                "ERROR at read_buf[%d]: expected 0x%02X, got 0x%02X\n", 
                i, write_buf[i], read_buf[i]);
	}
}

#define TEST_AREA_MAX DT_PROP(DT_NODELABEL(flash42), size)
#define EXPECTED_SIZE	1024
#define CANARY		0xff

static uint8_t __aligned(4) expected[EXPECTED_SIZE];

static void test_setup1(void)
{
	int rc;
	const struct device *flash_dev;
	struct flash_pages_info page_info;

	flash_dev = device_get_binding(FLASH_DEVICE);
	const struct flash_parameters *flash_parameters =
			flash_get_parameters(flash_dev);

	/* For tests purposes use page (in nrf_qspi_nor page = 64 kB) */
	flash_get_page_info_by_offs(flash_dev, TEST_REGION_OFFSET,
				    &page_info);

	/* Check if test region is not empty */
	uint8_t buf[EXPECTED_SIZE];

	rc = flash_read(flash_dev, TEST_REGION_OFFSET,
			buf, EXPECTED_SIZE);
	zassert_equal(rc, 0, "Cannot read flash");

	/* Fill test buffer with random data */
	for (int i = 0; i < EXPECTED_SIZE; i++) {
		expected[i] = i;
	}

	/* Check if tested region fits in flash */
	zassert_true((TEST_REGION_OFFSET + EXPECTED_SIZE) < TEST_AREA_MAX,
		     "Test area exceeds flash size");

	/* Check if flash is cleared */
	bool is_buf_clear = true;

	for (off_t i = 0; i < EXPECTED_SIZE; i++) {
		if (buf[i] != flash_parameters->erase_value) {
			is_buf_clear = false;
			break;
		}
	}

	if (!is_buf_clear) {
		/* erase page */
		rc = flash_erase(flash_dev, page_info.start_offset,
				 page_info.size);
		zassert_equal(rc, 0, "Flash memory not properly erased");
	}

}

static void test_read_unaligned_address(void)
{
	int rc;
	uint8_t buf[EXPECTED_SIZE];
	const struct device *flash_dev;
	struct flash_pages_info page_info;

	flash_dev = device_get_binding(FLASH_DEVICE);

	flash_get_page_info_by_offs(flash_dev, TEST_REGION_OFFSET,
				    &page_info);

	rc = flash_write(flash_dev,
			 page_info.start_offset,
			 expected, EXPECTED_SIZE);
	zassert_equal(rc, 0, "Cannot write to flash");

	/* read buffer length*/
	for (off_t len = 0; len < 25; len++) {
		/* address offset */
		for (off_t ad_o = 0; ad_o < 4; ad_o++) {
			/* buffer offset; leave space for buffer guard */
			for (off_t buf_o = 1; buf_o < 5; buf_o++) {
				/* buffer overflow protection */
				buf[buf_o - 1] = CANARY;
				buf[buf_o + len] = CANARY;
				memset(buf + buf_o, 0, len);
				rc = flash_read(flash_dev,
						page_info.start_offset + ad_o,
						buf + buf_o, len);
				zassert_equal(rc, 0, "Cannot read flash");
				zassert_equal(memcmp(buf + buf_o,
						expected + ad_o,
						len),
					0, "Flash read failed at len=%d, "
					"ad_o=%d, buf_o=%d", len, ad_o, buf_o);
				/* check buffer guards */
				zassert_equal(buf[buf_o - 1], CANARY,
					"Buffer underflow at len=%d, "
					"ad_o=%d, buf_o=%d", len, ad_o, buf_o);
				zassert_equal(buf[buf_o + len], CANARY,
					"Buffer overflow at len=%d, "
					"ad_o=%d, buf_o=%d", len, ad_o, buf_o);
			}
		}
	}
}

static void test_low_power()
{
	const struct device *flash_dev;
    int err;
	flash_dev = device_get_binding(FLASH_DEVICE);

#if IS_ENABLED(CONFIG_DEVICE_POWER_MANAGEMENT)
	printk("Putting the flash device into low power state...\n");
	err = device_set_power_state(flash_dev, DEVICE_PM_LOW_POWER_STATE,
				     NULL, NULL);
    zassert_equal(err, 0, "Setting low power mode failed");
#endif
}


void test_main(void)
{
    ztest_test_suite(flash_driver_test,
            ztest_unit_test(test_get_binding),
            ztest_unit_test(test_erase_read_write),
            ztest_unit_test(test_full_erase_full_read_write),
            ztest_unit_test(test_setup1),
            ztest_unit_test(test_read_unaligned_address),
            ztest_unit_test(test_low_power)
            );

    ztest_run_test_suite(flash_driver_test);
}
